packages/instant/src/components/timed_progress_bar.tsx

```diff
@@ -0,0 +1,102 @@
+import * as _ from 'lodash';
+import * as React from 'react';
+
+import { PROGRESS_FINISH_ANIMATION_TIME_MS, PROGRESS_STALL_AT_PERCENTAGE } from '../constants';
+import { ColorOption, keyframes, styled } from '../style/theme';
+
+import { Container } from './ui/container';
+
+export interface TimedProgressBarProps {
+    expectedTimeMs: number;
+    ended: boolean;
+}
+
+interface TimedProgressBarState {
```

Comment:
Do you need state? Seems like everything can be computed from `props`.
Something like (pseudoCode):
```
interface AnimationSettings / TimedProgressProps { 
  animationTimeMs: number;
  animationStartingWidth: string;
  maxWidthPercent: number;
}

private readonly _calculateAnimationSettings(): AnimationSettings => {
  if (this.props.isEnded) {
    return {
      animationTimeMs: props.endingAnimationTime,
      animationStartingWidth: this._readWidthFromDOM(),
      maxWidthPercent: 100,
    }
  }
  return {
    animationTimeMs: props.estimatedTime;
    animationStartingWidth: 0;
    maxWidthPercent: props.maxWidthToStallAtPercent;
  } 
}

public render(): React.ReactNode {
  return <ProgressBar {{...this._calculateAnimationSettings()}}/>
}

```
I think this is basically what you have but you're saving to state.

Also would it be possible to have start width and end width use the same units?
Don't think there should be both `TimedProgressProps` and `AnimationSettings` or whatever it ends up being.

Reply:
I was unsure about how grabbing the ref width would work outside of the `componentDidUpdate` callback, but it turns out you are so right: I didn't need state here, and it is much cleaner now without it  💯

You deserve a guest post on https://you-dont-need.com/ :)

Reply:
Lol our code is not a good example of "you-dont-need"